src/log_manager/log_manager_mutex_impl.h

```diff
@@ -1,30 +1,267 @@
+                   entry.data_.length());
+            input_data.append(data);
+            offset_and_term_.push_back(std::make_pair(file_size, entry.term_));
+            file_size += ENTRY_HEAD_SIZE + entry.data_.length();
+        }
+        fd_.write(input_data.c_str(), input_data.length());
+        if (will_sync) fd_.sync();
+        return 0;
+    }
+
+    // get entry by index
+    LogEntry get_LogEntry(const int64_t index) {
+        std::unique_lock<std::mutex> lock(file_mutex_);
+        LogEntry entry;
+        if (index > offset_and_term_.size() || index <= 0) return entry;
+        fd_.seekg(offset_and_term_[index - 1].first);
+        char headBuf[ENTRY_HEAD_SIZE];
+        memset(headBuf, 0, ENTRY_HEAD_SIZE);
+        fd_.read(headBuf, ENTRY_HEAD_SIZE);
+        EntryHeader header;
+        LogEntryPackage pack;
+        pack.unpackHeader(headBuf, header);
+        char *dataBuf = new char[header.data_len];
+        memset(dataBuf, 0, header.data_len);
+        fd_.read(dataBuf, header.data_len);
+        uint32_t sum = 0;
+        for (size_t i = 0; i < header.data_len; i++) {
+            sum = sum ^ dataBuf[i];
+        }
+        if (sum != header.data_checksum) {
+            // TODO: abort()?
+            std::cout << "check sum invalid." << std::endl;
+        }
+
+        entry.term_ = header.term;
+        entry.index_ = index;
+        entry.data_ = dataBuf;
+        entry.data_.resize(header.data_len);
+        delete[] dataBuf;
+
+        return entry;
+    }
+
+    // get entry's term by index
+    int64_t get_term(const int64_t index) const {
+        if (index > offset_and_term_.size() || index < 0) return -1;
+        return offset_and_term_[index - 1].second;
+    }
+    int64_t get_size(const int64_t index) const {
+        if (index > offset_and_term_.size() || index < 0) return -1;
+        return offset_and_term_[index - 1].first;
+    }
+    int64_t get_count() const { return offset_and_term_.size(); }
+    // truncate data to last_index_kept
+    int truncate(const int64_t last_index_kept) {
+        std::unique_lock<std::mutex> lock(file_mutex_);
+        if (last_index_kept >= offset_and_term_.size()) {
+            return 0;
+        }
+        if (last_index_kept < 0) return -1;
+        std::filesystem::resize_file(path_, offset_and_term_[last_index_kept].first);
+        //        fd_.seekg(0, std::ios::end);
+        if (last_index_kept == offset_and_term_.size() - 1)
+            offset_and_term_.pop_back();
+        else {
+            int64_t diff = offset_and_term_.size() - last_index_kept;
+            offset_and_term_.erase(offset_and_term_.end() - diff, offset_and_term_.end());
+        }
+        return 0;
+    }
+
 private:
     std::mutex queue_mutex_;
     std::condition_variable queue_cv_;
     std::queue<LogEntryType> queue_;
+
+    std::mutex file_mutex_;
+    std::string path_;
+    bool is_open_;
+    bool is_init_;
+    std::fstream fd_;
```

Comment:
Why we need fstream? It's better to implement the in-memory log manager first, so there's no need to write data to any file.